### PR TITLE
new `macros.genAst`: sidesteps issues with `quote do`

### PR DESCRIPTION
**Pull request by timotheecour**
provides a solution for https://github.com/nim-lang/RFCs/issues/122

a new `macros.genAst` is introduced; it sidesteps virtually all issues inherent with `quote do` and can be used as a replacement for it
```nim
macro fun(a: bool, b: static bool): untyped =
  let b2 = newLit b
  let c = newLit true
  quote do: echo (`a`, `b2`, `c`)

# can be replaced by:
macro fun(a: bool, b: static bool): untyped =
  genAst(a, b, c = true): echo (a, b, c)
```

## sidesteps these (see unittests tests/macros/tgenast.nim)
* #7375
* #8220 via `getAstOpt({kDirtyTemplate})` see unittests
* #7589
* #7726
* #9607
* #11986
* EDIT: plus other issues, eg recursive use, ambiguity with backticks (eg operators that need backticks don't work with `quote`)

for #7889, see unittests showing this is fixed (for both `quote do` and `genAst`) by using regular call syntax instead of method call syntax; can be therefore closed as dup of https://github.com/nim-lang/Nim/issues/7085

## comparison to `quoteAst` proposal
this PR doesn't have limitations of the other `quoteAst` proposal mentioned in https://github.com/nim-lang/RFCs/issues/122#issue-401636535 : see unittest `tests/macros/tgenast.nim` (`macro abc(name: untyped)` test case)
> As it seems, the uq(...) expression is not allowed everywhere. The following example does not pass the parsing step of Nim, because an unquote expression is now allowed as the name of a new type

it doesn't require any parser change unlike what's mentioned in https://github.com/nim-lang/RFCs/issues/122 
> To fix this problem the parser needs to be changes.

it's also much simpler than https://github.com/nim-lang/Nim/pull/10446 (no compiler change needed, just 1 new proc in macros.nim)

so I think this PR also closes https://github.com/nim-lang/RFCs/issues/122 and https://github.com/nim-lang/Nim/pull/10446 

## note
* the only issue with `quote do` that is not fixed by this PR is https://github.com/nim-lang/Nim/issues/10430 `Comments are removed in quote do expressions` , however this is in fact a pre-existing problem with `getAst` that can be easily fixed independently of this PR (possibly via an extra argument to `getAst`)

* there are use cases for when `dirty` pragma is desirable, so there is an option to control that:`getAstOpt({kDirtyTemplate}, ...)` , see unittests. However, this leads to more verbose code (each symbol used must be in capture list) and potentially surprising hijacking behavior in case some symbols are not captured (see unittests), therefore the default is `not dirty`. See nim docs for what symbols are gensym'd vs inject'd: https://nim-lang.github.io/Nim/manual.html#templates-hygiene-in-templates

### CODE REVIEW

**Comment by Araq** [comment]
Looks *very* nice, but @krux02 needs to do the final review and he is on holidays. ;-)

**Comment by kaushalmodi** [comment]
@timotheecour Can you also add some more "readme" in the macro tutorial (tut 3) to compare this to `quote do`?

Thanks.

**Comment by timotheecour** [comment]
@kaushalmodi 
> Can you also add some more "readme" in the macro tutorial (tut 3) to compare this to quote do?

could be done in subsequent PR

File: lib/core/macros.nim [hunk 1]
```diff
@@ -1344,6 +1348,93 @@ proc copy*(node: NimNode): NimNode {.compileTime.} =
...
+    case a.kind
+    of nnkExprEqExpr:
+      varName = a[0]
+      varVal = a[1]
+    of nnkIdent:
+      varName = a
+      varVal = a
+    else: error("invalid argument kind: " & $a.kind, a)
+
+    params.add newTree(nnkIdentDefs, [varName, newEmptyNode(), newEmptyNode()])
```
Comment (krux02) [inline lib/core/macros.nim]: NewTree takes ``varargs[NimNode]``, not ``openArray[NimNode]``. I am afraid, but scheduled fixes for varargs will trigger a compiler error here in the future, because ``[varName, newEmptyNode(), newEmptyNode()]`` is not a ``NimNode``, but an array of ``NimNode``.

File: lib/core/macros.nim [hunk 2]
```diff
@@ -1344,6 +1348,93 @@ proc copy*(node: NimNode): NimNode {.compileTime.} =
   ## An alias for copyNimTree().
   return node.copyNimTree()
 
+type GenAstOpt* = enum
+  kNoExposeLocalInjects,
```
Comment (krux02) [inline lib/core/macros.nim]: I am sorry, but you have to find a better name for that enum, this is just cryptic.

File: lib/core/macros.nim [hunk 3]
```diff
@@ -1344,6 +1348,93 @@ proc copy*(node: NimNode): NimNode {.compileTime.} =
...
+    # when unset, inject'd symbols (including implicit ones such as local procs
+    # in scope) are exposed implicitly;
+    # gensym'd symbols will generate a CT internal error: `environment misses:`
+    # when set, local symbols are not exposed unless captured explicitly in
+    # `genAst` argument list. The default is unset, to avoid surprising hijacking
+    # of local symbols by symbols in caller scope.
```
Comment (krux02) [inline lib/core/macros.nim]: This comment should be rephrased. It is really hard for me to read and understand it.

File: lib/core/macros.nim [hunk 4]
```diff
@@ -1344,6 +1348,93 @@ proc copy*(node: NimNode): NimNode {.compileTime.} =
...
+        # echo xignored # would give: Error: undeclared identifier
+        # echo s0 # would give: Error: internal error: expr: var not init s0_237159
+        (s1, s2, x0, x1, x2, x3, localExposed())
+
+    macro bar2(x0: static Foo, x1: Foo, x2: Foo, xignored: Foo): untyped =
+      let s0 = "not captured!" ## does not override `s0` from caller scope
+      let s1 = "not captured!" ## does not override `s1=2`
+      let xignoredLocal = kfoo4
+      let x3 = newLit kfoo4
+      result = genAst({kNoExposeLocalInjects}, s1=2, s2="asdf", x0=newLit x0, x1=x1, x2, x3):
```
Comment (krux02) [inline lib/core/macros.nim]: s2 is a string, not a literal. This should be a compilation error. Please enforce the usage of ``newLit`` to make clear that only NimNodes can be captured.

File: lib/core/macros.nim [hunk 5]
```diff
@@ -1344,6 +1348,93 @@ proc copy*(node: NimNode): NimNode {.compileTime.} =
...
+type GenAstOpt* = enum
+  kNoExposeLocalInjects,
+    # when unset, inject'd symbols (including implicit ones such as local procs
+    # in scope) are exposed implicitly;
+    # gensym'd symbols will generate a CT internal error: `environment misses:`
+    # when set, local symbols are not exposed unless captured explicitly in
+    # `genAst` argument list. The default is unset, to avoid surprising hijacking
+    # of local symbols by symbols in caller scope.
+
+macro genAst*(options: static set[GenAstOpt] = {}, args: varargs[untyped]): untyped =
```
Comment (krux02) [inline lib/core/macros.nim]: I am not a fan of this ``options`` argument at the beginning. AFAIK the default argument doesn't work here. So I always have to write the curly pair for genAst. Can't we get rid of these flags argument entirely, for example by providing two different names for genAst for the two different behaviors.

**Comment by zah** [comment]
This looks very nice indeed.

Would you please add a test for a nested application of `genAst`? `quote` is currently not capable of generating the body of another macro that will in turn also use `quote`. Here is one synthetic example:

```nim
macro createMacro(name, obj, field: untyped): untyped =
  result = genAst({}, obj = newDotExpr(obj, field), lit = newLit(10)):
    macro name(arg: untyped): untyped =
      result = genAst({}):
        echo obj
        echo astToStr(field)
        echo lit
        echo arg

var x = @[1, 2, 3]
createMacro foo, x, len
foo 20
```

The expected output is:

```
3
len
10
20
```

**Comment by timotheecour** [comment]
@zah just added your test case, i had to make minor adjustments to make it work

haven't addressed other comments yet

**Comment by krux02** [comment]
I think I have to take it back. I am not impressed anymore.

Your claim to fix #10326 is a lie. That issue is already fixed. I tested it.
Your claim to fix #7375 is a lie. That bug exists with your genAst as well as with the old quote  do

```nim
import macros

macro fails(b: static[bool]): untyped =
  echo b
  result = newStmtList()

macro works(b: static[int]): untyped =
  echo b
  result = newStmtList()

macro foo(): untyped =

  var b = false

  ## Fails
  result = genAst({},b):
    fails(b)

  ## Works
  # result = quote do:
  #   works(`b`)

foo()
```
output
```
/tmp/scratch.nim(27, 4) template/generic instantiation of `foo` from here
/tmp/scratch.nim(20, 19) Error: type mismatch: got <static[int literal(0)](0)>
but expected one of: 
macro fails(b: static[bool]): untyped
  first type mismatch at position: 1
  required type for b: static[bool]
  but expression '0' is of type: static[int literal(0)](0)
```


Your claim to fix #9745 is a lie. That issue is already fixed.
I could not get #7889 to work with your genAst either.


I tried if #8220 works with your PR, and surprise I could get it working:

```nim
import macros, strformat

macro foo(): untyped =
  result = genAst({kNoExposeLocalInjects}):
    let bar = "Hello, World"
    echo &"Let's interpolate {bar} in the string"

foo()
```

But still, the reasoning why this works with ``{kNoExposeLocalInjects}``, but not with ``{}`` is incomprehensible. 

Your claim to fix #7589, nope. Could not get it to work.

About your claim to fix #7726, I have to give you that, there is a solution for it:

```nim
import macros

macro foo(): untyped =
  let a = @[1, 2, 3, 4, 5]
  result = genAst({}, len = a.len):
    len

echo foo() # 5
```

But this solution is not structurally safer than the already exiting solution with quote do:

```nim
macro foo(): untyped =
  let a = @[1, 2, 3, 4, 5]
  let len = a.len
  result = quote do:
    `len`
```

Your solution still causes interal errors in the compiler if you forget to properly escape the arguments:

```nim
import macros

macro foo(): untyped =
  let a = @[1, 2, 3, 4, 5]
  result = genAst({}):
    a.len  # # Error: internal error: expr: var not init a_186029

echo foo() 
```

That example also represents more closely the problem from the original bug. So sorry I can't approve your PR here as a fix. A fix would provide here a proper error message that complains about an unknown identifier `a`.

Now about your own issue: #9607
When I change that example to use your ``genAst``, it is still plagued by internal errors:

```nim
import macros

proc fun1*(info:LineInfo): string = "bar"
proc fun2*(info:int): string = "bar"

macro echoL*(args: varargs[untyped]): untyped =
  let info = args.lineInfoObj
  let fun1 = bindSym"fun1"
  let fun2 = bindSym"fun2"

  result = genAst({}, fun1, info):  # 
    fun1(info)

echo echoL()
```

```Error: internal error: getTypeDescAux(tyUntyped)```


I had enough

**Comment by bluenote10** [comment]
@krux02 Why do you close the PR instead of asking for working on the open points? Such a review style is not making the Nim community a big favor.

**Comment by kaushalmodi** [comment]
@krux02 The way you worded your remarks ("your claim is a lie", "i had enough", etc.) reeks of some sort of personal aggression you have against @timotheecour (I am saying this because this is not the first time I am seeing this from you).

I have in no position to validate this PR, but looking at the amount of code and time @timotheecour could have spent on creating this, I don't think your reaction justifies this.

Such angry PR closures will hurt/are hurting the Nim community.

**Comment by dom96** [comment]
My initial reaction was the same as @bluenote10's and @kaushalmodi's, but after reading the PR I am more sympathetic to @krux02's frustration. From my side I have tried multiple times to understand this `genAst` macro and I still do not understand how I would use it instead of ``quote``, I think that @timotheecour should consider showing simple examples first, not going straight to super complex ones. This is especially true for the ``runnableExamples`` he's added to this macro, the fact that they look more indecipherable than the actual implementation of the ``genAst`` macro is telling.

Even so, this could have been handled much better, this PR is dense and those checking out Nim are likely to pick out the phrases that @kaushalmodi picked out above and get a negative opinion about Nim's community. In addition, many contributors to Nim are volunteering their time and often work very hard to get something accepted into the language. Someone who is so active in Nim development (and even more importantly, somebody who's employed to work on Nim) would do well to use more positive and encouraging language.

Remember that when you speak to a contributor, you are not just speaking to that contributor, you are speaking to anyone who is checking out the community.

File: lib/core/macros.nim (hunk 5 above)
Comment (timotheecour) [reply lib/core/macros.nim]: I fixed it in latest commit (doens't appear here since you closed this PR)

File: lib/core/macros.nim (hunk 4 above)
Comment (timotheecour) [reply lib/core/macros.nim]: I fixed it in latest commit (ditto, doesn't appear since you closed PR); `newLit` is now done automatically if needed

File: lib/core/macros.nim (hunk 1 above)
Comment (timotheecour) [reply lib/core/macros.nim]: fixed in latest commit

File: lib/core/macros.nim (hunk 2 above)
Comment (timotheecour) [reply lib/core/macros.nim]: fixed in latest commit

File: lib/core/macros.nim (hunk 3 above)
Comment (timotheecour) [reply lib/core/macros.nim]: improved it in latest commit

**Comment by timotheecour** [comment]
The common etiquette on github is to let PR author reply to review comments before closing his PR. Closing my issues and PR's immediately after leaving a review comment isn't fostering collaboration. This keeps happening, eg my latest PR https://github.com/nim-lang/Nim/pull/11732 was the same; someone else reopened / merged it after comments were addressed.
Blocking me on github (https://github.com/nim-lang/RFCs/issues/122 still shows as "You can't perform this action at this time") also isn't fostering collaboration.

Back to the PR:

I've pushed new commits that hopefully address all comments + concerns. Unfortunately my PR was closed by @krux02 so the new commits don't appear here.

With latest commit, 
* #9607 is indeed fixed (newLit not needed anymore, no internal errors, etc)
* #7726 is indeed fixed, the point is things like "`a.len`" don't work in quote do and with getAst you don't need to pollute outer scope with `let len = newLit a.len`, in fact with latest commit you can do:
```nim
getAst(len=a.len): echo len
```
* #7589 is indeed fixed, see test suite in this PR `tests/macros/tgenast`: `quote do` can't deal with ambiguity of symbols that need quoting such as operators "`==`", getAst does not have that issue
* #8220 also is in test suite and  reason for `{kNoExposeLocalInjects}` is explained in latest commit
* #7375: as mentioned above, the point is that you don't need to pollute outer scope with a newLit declaration, it's done in argument list; 
and in fact in latest commit newLit is not needed anymore, it's auto added when needed so that 
```nim
result = genAst(b): fails(b)` # works
result = quote do: fails(`b`) # fails CT error
```
which fixes #7375

* #7889 : it's in the test suite; I had to change method call syntax to make it work, but that's an inherent limitation with `templates`, which already has a github issue for it. So I considered this issue fixed. Although I've just tried it today with `quote do` and using method call syntax also made it work, so we can probably just close #7889 as duplicate of template + method call syntax issue.

As for the last 2, #10326 and #9745, I had interpreted it in the more general context of backticks causing ambiguity, see example I had for that in test suite which failed with `quote do` but works with getAst:
```nim
result = quote do: # will fail
    template `a1=`(x: var Foo, val: int) =
        x.a = val
```
but indeed, you can consider that issue as fixed prior to this PR as this example is of more general scope than the issue of backticks in pragmas


> I still do not understand how I would use it instead of quote, I think that @timotheecour should consider showing simple examples first, not going straight to super complex ones.

I've hopefully addressed/clarified this in latest commit, eg:
```nim
macro fun(a: bool, b: static bool): untyped =
  let b2 = newLit b
  quote do: echo (`a`, `b2`)
```
can be replaced by:
```nim
macro fun(a: bool, b: static bool): untyped =
  genAst(a, b): echo (a, b) # newLit auto applied when needed
```

**Comment by krux02** [comment]
@timotheecour what is bothering me the most are the false claims, not the technical deficiencies. Remove those false claims and apologize for bringing them up in the first place. Then we can talk about the technical deficiencies.

It is the same with your other PR [make sets (tables etc) 1000 times faster for pointer, int64, int etc]( https://github.com/nim-lang/Nim/pull/11767). You just promise everything from the world. But you don't explain the technical deficiencies in the old hash and how you addressed them. Also you didn't bother to mention the shortcomings of your algorithm (string allocations for everything). If you can't prove/show/explain why your way of doing something is or should be better, then please don't claim that it is better.

**Comment by zah** [comment]
I think closing this issue was a bit immature on Krux's side, but more importantly it was also objectively irrational. 

Our personal feelings should not get in the way of making Nim a better language and it doesn't take much knowledge of the compiler internals to realize that @timotheecour has invented a promising new way to sidestep many of the foundational issues with the existing quasi-quoting approach. 

Even if the initial implementation is not perfect, the direction is worth pursuing and we should be excited by this new development. We should thank Timothee for being passionate about Nim, for spending so much time on it, and for coming with all these new ideas.

-----

@timotheecour, you should also add a flag controlling the dirtiness of the generated template. The conclusions from previous discussions were that `dirty` is a better default for quasi-quoting.

**Comment by timotheecour** [comment]
@zah
> you should also add a flag controlling the dirtiness of the generated template.

that's what the flag `kNoExposeLocalInjects` is about (which I renamed `kDirtyTemplate` in latest commit, which, again, won't appear here unless PR is reopened)

> The conclusions from previous discussions were that dirty is a better default for quasi-quoting.

that was what i tried first, but after trying it out it became clear that non-dirty was a better default to avoid surprising hijacking behavior; dirty is occasionally useful though, see unittests in tests/macros/tgenast.nim 

@krux02

I believe I've answered that here https://github.com/nim-lang/Nim/pull/11722#issuecomment-514456533

> It is the same with your other PR make sets (tables etc) 1000 times faster for pointer, int64, int etc.
> But you don't explain the technical deficiencies in the old hash and how you addressed them. 

you've already asked me that here https://github.com/nim-lang/Nim/pull/11767#issuecomment-512604271 and I've already answered it https://github.com/nim-lang/Nim/pull/11767#issuecomment-512611465 ; happy to provide more clarifications if needed

> Also you didn't bother to mention the shortcomings of your algorithm (string allocations for everything). If you can't prove/show/explain why your way of doing something is or should be better, then please don't claim that it is better.

my claim was backed by benchmarking code I've shared in the original post of that PR; anyone is welcome to verify/criticize/improve it; that benchmark showed the string alloc's cost that you mention were dwarfed by the other costs caused by collisions in the pre-existing hashing scheme

**Comment by zah** [comment]
> surprising hijacking behavior

Can you clarify this problem with some examples?

**Comment by krux02** [comment]
> @krux02
> 
> I believe I've answered that here [#11722 (comment)](https://github.com/nim-lang/Nim/pull/11722#issuecomment-514456533)

Your false claims are still in your part of your pull request description. They need to be edited out. Then I also asked to apologise, I could not find anything like that.

**Comment by timotheecour** [comment]
@zah
> Can you clarify this problem with some examples?

* see example `bindme5UseExposeFalse` in the PR https://github.com/nim-lang/Nim/pull/11722/files#diff-f0de32b58ad2d14a98d80be161ed74f0R106, paraphrased below (and using syntax from latest commit):

```nim
# ./mgenast.nim
proc gun(): auto = false
macro mgun*(): untyped =
  genAstOpt({kDirtyTemplate}): gun()
  # library writer probably intended this to call `mgenast.gun`,
  # but that symbol is hijacked by the one in caller scope, eg `mylib.gun`

# ./main.nim
import./mylib # defines `proc gun(): auto = true`
# or just directly defining `gun` locally
import./mgenast
echo mgun() # true
```

* the 2nd problem with `kDirtyTemplate` as being the default is it leads to more verbose code in practice, where library writer needs to specify every single symbol referenced in the capture list, see example I had for that in the PR ```nim
macro bindme6UseExposeFalse*(): untyped =
  ## with `kDirtyTemplate`, requires passing all referenced symbols which can be tedious
  genAstOpt({kDirtyTemplate}, newStringStream, writeData, readData):
    var tst = "sometext"
    var ss = newStringStream("anothertext")
    writeData(ss, tst[0].addr, 2)
    discard readData(ss, tst[0].addr, 2)
```

**Comment by zah** [comment]
Well, both of these are completely expected behavior. You can also use bind statements as a work-around. I think Nim will eventually have to gain support for top-level mixin/bind statements, because otherwise you cannot control the binding behavior inside type sections. They will make it slightly easier to use the dirty version of `genAst` as well.
 
I guess we need a little bit more experience to decide what is the better default. I would advocate for a short period where the default isn't set, so we can gain more experience with the API.

**Comment by timotheecour** [comment]
@krux02
> Your false claims are still in your part of your pull request description. They need to be edited out. Then I also asked to apologise, I could not find anything like that.

I've edited the PR description to reflect the state of this PR as of the latest commit I pushed (which doesn't appear here because this PR is closed), to the best of my understanding. If anything is still inaccurate as of latest commit, I will happily correct it. 
So I've crossed out #10326 #9745 which are indeed already fixed; note: they still appear in the fix list in your open PR https://github.com/nim-lang/Nim/pull/10446:
```
fixed issues
This is the list of issues that will be fixed when this PR is finished: #10326 #7375 #9745 #7889 #8220 #7589 #7726 #10430
```

Alternatively, I can also remove any mention of fixed issues from this PR and defer to someone else the task of figuring out which issues can be closed.

I hope the new PR you've opened https://github.com/nim-lang/Nim/pull/11823 after closing my PR will materialize soon so we can finally fix `quote do`, unlike previous instance where you closed my PR (that would've helped exposing nim comments to user code) in favor of your PR but never finished it, eg https://github.com/nim-lang/Nim/pull/10054#issuecomment-451815745

**Comment by krux02** [comment]
I am reopening this PR now. But I am still not happy with it. 

There is a lot in your wording that bothers me a lot. The issue is still titled ``new `macros.genAst`: fixes all issues with `quote do` ``. No it does not fix *all* issues with quote do. The issue with ``quote do`` are still there, becaues you din't fix anything in `quote do`. You just provide an experimental alternative to `quote do` that does not have the *reported* issues of `quote do`. Then I asked you to edit out the issues you did not fix ... and you leave them in. Striked though, but still you leave them in. Why? This makes the PR overview harder to read. The overview should be as concise as possible. Only mention what you really did and improve. If you can remove words, remove them, don't cross them out. Nobody who reviews a PR wants to see if it once had false claims in the description.

For the comment field issue, I am sorry. My branch is indeed on ice. My approach did not work out eventually. I will reopen your PR as it becomes relevant again. 

And sorry abot being blocked in my PRs. I still did not unblock your personal account.

**Comment by timotheecour** [comment]
PTAL, thanks for reopening.

* changed PR title, description, and PR content to use `sidestep` instead of `fix`, + other comment improvements
* simplified runnableExamples

> No it does not fix all issues with quote do. The issue with quote do are still there, becaues you din't fix anything in quote do. You just provide an experimental alternative to quote do that does not have the reported issues of quote do

I'm not sure how all of `quote`'s issues can be fixed in a backward compatible way, so I don't see the value in keeping those linked issues opened if there's an alternative (eg `genAst`) that doesn't have those issues and that can be used as replacement for `quote`

> Striked though, but still you leave them in. Why? 

I've now removed them. the reason I had striked them was to avoid confusion for ppl reading through comment history

**Comment by krux02** [comment]
> I've now removed them. the reason I had striked them was to avoid confusion for ppl reading through comment history

Don't do that. you just create confusion for people reading the issue description. The issue has a drop down menue ``edited``, people who want to see the original unedited post can do that there.

File: tests/macros/tgenast.nim [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+
+block:
+  macro bar(x0: static Foo, x1: Foo, x2: Foo, xignored: Foo): untyped =
+    let s0 = "not captured!"
+    let s1 = "not captured!"
+    let xignoredLocal = kfoo4
+
+    # newLit optional:
+    let x3 = newLit kfoo4
+    let x3b = kfoo4
```
Comment (krux02) [inline tests/macros/tgenast.nim:13]: I am sorry, I have to say no to automatic `newLit` injection. I this this is one of the major problem sources of ``quote do`` and you ported this problem source to ``genAst``. There should be an error message when you try to capture anything that is not of type ``NimNode``.

Comment (timotheecour) [reply tests/macros/tgenast.nim:13]: `quote do` doesn't do automatic `newLit` injection, and suffers from bugs like https://github.com/nim-lang/Nim/issues/7375.
I haven't ported this problem to `genAst`, IMO it's the opposite: automatic `newLit` injection as done in this PR not only avoids this category of bugs, but it also simplifies usage, avoiding the redundant `x = newLit x` for every non-NimNode parameter:
`genAst(x1, x2, x3 = true): body`
instead of:
`genAst(x1 = newLit(x1), x2 = newLit(x2), x3 = newLit(true)): body` which gets tiring after a while

Note, the `newLit` injection only happens if the type is not `NimNode`, and the user can always provide his own transformation instead, eg:
`genAst(x1 = myNewLit(x1)): body` and it'll do the right thing

Lastly, there's always `kNoAutoNewLit` in case user wants to avoid automatic `newLit` injection for some reason.

Comment (krux02) [reply tests/macros/tgenast.nim:13]: I think it is logically wrong to treat `NimNode` and non-`NimNode` differently. Either everything gets newLit applied, but then also NimNode, where expressions are generated to create NimNode values, or nothing. This attitude towards "automatically doing the right thing" usually ends up in "automatically doing the wrong thing". Another question is, what do you do with function symbols? What do you do with type symbols?

Comment (krux02) [reply tests/macros/tgenast.nim:13]: And btw ``kNoAutoNewLit`` is also yet another crappy flag. Sorry I hate super verbose flags that I have to add to everything. ``genAst`` should not have a flags argument.

Comment (timotheecour) [reply tests/macros/tgenast.nim:13]: > genAst should not have a flags argument.

it doesn't, only `genAstOpt` has a flags argument; most code will just use `genAst(x,y): body` and not bother with flags

> And btw kNoAutoNewLit is also yet another crappy flag. Sorry I hate super verbose flags that I have to add to everything

renamed it to `kNoNewLit`, suggestions for alternate names welcome. This flag should very rarely be used so most users will most likely not have to use it


> Either everything gets newLit applied,

that can't work

> or nothing

that'll result in more verbose code littered with `newLit` as discussed above, and the only sane way to do that would be to error when trying to capture anything that is not of type NimNode, however, see point below:

> There should be an error message when you try to capture anything that is not of type NimNode

that would break passing proc symbols, see `bindme6UseExposeFalse` example in mgenast.nim for example:
```nim
macro bindme6UseExposeFalse*(): untyped =
  genAstOpt({kDirtyTemplate}, newStringStream, writeData, readData):
    var tst = "sometext"
    var ss = newStringStream("anothertext")
    writeData(ss, tst[0].addr, 2)
    discard readData(ss, tst[0].addr, 2)
```

> This attitude towards "automatically doing the right thing" usually ends up in "automatically doing the wrong thing"

do you have a realistic concrete example where automatic newLit injection does the wrong thing and an alternative implementation does the correct thing?

> Another question is, what do you do with function symbols? What do you do with type symbols?

function symbols can be captured explicitly, but the capture is optional because of gensym rules (https://nim-lang.github.io/Nim/manual.html#templates-hygiene-in-templates), see `test passing function and type symbols` in tgenast.nim; when `kDirtyTemplate` is used, they must be captured (see `bindme6UseExposeFalse` for an example); it works despite the type not being NimNode, as mentioned above.

type symbols can cause a `Error: internal error: expr(nkType); unknown node kind`, with or without automatic NewLit injection; so auto NewLit injection doesn't make anything worse.

Of course if the function or type symbol is actually of type NimNode (eg a param of the macro, or the result of a `genAst` call (eg
`let Z2 = genAst(): type(true)`), there is nothing special, the capture works as usual

see tgenast.nim `block: #test passing function and type symbols` for an example 


TLDR: the large majority of code will just use the default `genAst: body` or `genAst(x,y): body` for captures and not bother with any flags, resulting in code that's easy to read and write ; the flags are there as an escape when default behavior is not desired for some reason.

Comment (zah) [reply tests/macros/tgenast.nim:13]: Without experience, it's hard to know for sure, but the automatic `newLit` application sounds like a reasonable feature to me. You can compare it to `echo` which automatically calls the `$` op for you.

Comment (krux02) [reply tests/macros/tgenast.nim:13]: First of all, applying newLit to everything is possible. You just need to provide an overload of `newLit` for ``NimNode`` that will generate an ast that will when evaluated, generate the input ast. I did that im my branch. I just didn't call that ``newLit``, because I don't plan to export it (for now).

> do you have a realistic concrete example where automatic newLit injection does the wrong thing and an alternative implementation does the correct thing?

You don't get it. If I merge this PR, I have to make sure that it will work in the future. If I merge this PR it becomes my liability not yours. And want to make sure that there are no bugs introduce because of surprising behavior. So _you_ have to convince me that the automatic newLit injection with the special casing of ``NimNode`` arguments is the most reliable way to do this. And I am not convinced right now.

> type symbols can cause a Error: internal error: expr(nkType); unknown node kind, with or without automatic NewLit injection; so auto NewLit injection doesn't make anything worse.

Internal errors fall into the category of a compiler crash. So they are are bug that need to be addressed.

Comment (timotheecour) [reply tests/macros/tgenast.nim:13]: > So they are are bug that need to be addressed.

here's a minimimal test case (independent of this PR), I can create a github issue if needed

```nim
import macros
macro bar(): untyped =
  template baz(T) = echo $T
  result = getAst(baz(int)) # Error: internal error: expr(nkType); unknown node kind
bar()
```

> There should be an error message when you try to capture anything that is not of type NimNode
> First of all, applying newLit to everything is possible

how would that work in the general case? what about closures ?
This works fine despite `fun` not being a `NimNode` (this example works independent of this PR):

```nim
import macros

var count = 0
proc fun() = count.inc
macro bar(): untyped =
  template baz(f) = f()
  result = getAst(baz(fun))
bar()
echo count
```

> And want to make sure that there are no bugs introduce because of surprising behavior

that could be said of any piece of software. I can't prove that this PR is bug free, but I did provide a not-so-small test suite (much more comprehensive than the one for `quote do`) and a fall-back option. If someone runs into a particular scenario where NewLit injection results in error, he can just use `{kNoNewLit}`, and maybe even add a `macros.genAstRaw` template to avoid passing the flag if that becomes unbearable for some reason.

File: lib/core/macros.nim [hunk 7]
```diff
@@ -1344,6 +1348,76 @@ proc copy*(node: NimNode): NimNode {.compileTime.} =
...
+  let params = newTree(nnkFormalParams, newEmptyNode())
+  let pragmas =
+    if kDirtyTemplate in options:
+      nnkPragma.newTree(ident"dirty")
+    else:
+      newEmptyNode()
+
+  template newLitMaybe(a): untyped =
+    when type(a) is NimNode: a
+    elif compiles(newLit(a)): newLit(a)
```
Comment (krux02) [inline lib/core/macros.nim]: Depending on ``compiles`` is very harmful. Sometimes the compiler becomes stricter and prints error messages. ``compilers`` eats them and secretly changes behavior.

File: lib/core/macros.nim [hunk 8]
```diff
@@ -1344,6 +1348,76 @@ proc copy*(node: NimNode): NimNode {.compileTime.} =
...
+
+  let params = newTree(nnkFormalParams, newEmptyNode())
+  let pragmas =
+    if kDirtyTemplate in options:
+      nnkPragma.newTree(ident"dirty")
+    else:
+      newEmptyNode()
+
+  template newLitMaybe(a): untyped =
+    when type(a) is NimNode: a
```
Comment (krux02) [inline lib/core/macros.nim]: It doesn't print a warning yet, but it will. ``typeof`` is the correct keyword to use here.

Comment (timotheecour) [reply lib/core/macros.nim]: done

File: lib/core/macros.nim (hunk 7 above)
Comment (timotheecour) [reply lib/core/macros.nim]: done; found a better way without `compiles`;
(makes me think would be nice to have a `compiles(foo(args), sigcheckOnly)` which would just return whether a proper overload can be found, without compiling body; this would also be more efficient and resolve a number of issues with `compiles`)

**Comment by timotheecour** [comment]
@krux02  PTAL. in particular, look at the example I added `mixinExample`  which shows how user can control precisely what symbols are to be resolved in caller scope vs callee scope. This IMO greatly reduces the need for dirty templates. The syntax stays minimal:
```nim
proc locafun1(): auto = "in locafun1"
proc locafun2(): auto = "in locafun2"
# locafun3 in caller scope only
macro mixinExample*(): untyped =
  genAst:
    mixin locafun1
    (locafun1(), locafun2(), locafun3())
```

File: lib/core/macros.nim (hunk 7 above)
Comment (krux02) [reply lib/core/macros.nim]: we actually think about something like that internally. Something like ``overloadExists(foo(args))``.

Comment (timotheecour) [reply lib/core/macros.nim]: => moved that discussion to https://github.com/nim-lang/RFCs/issues/164

**Comment by Araq** [comment]
Please add this implementation not directly to ``macros.nim`` but to some new ``std / macrogenerators`` module. And create a new PR for it so that we can have a civilized discussion about it this time.

**Comment by timotheecour** [comment]
@Araq done, see https://github.com/nim-lang/Nim/pull/17426

**Comment by timotheecour** [comment]
finally merged! (see https://github.com/nim-lang/Nim/pull/17426)
